src/tss2-tcti/tcti-swtpm.c

```diff
@@ -0,0 +1,632 @@
+    TSS2_RC rc;
+
+    if (tcti_swtpm == NULL) {
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+    rc = tcti_common_transmit_checks (tcti_common, cmd_buf);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+    rc = header_unmarshal (cmd_buf, &header);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+    if (header.size != size) {
+        LOG_ERROR ("Buffer size parameter: %zu, and TPM2 command header size "
+                   "field: %" PRIu32 " disagree.", size, header.size);
+        return TSS2_TCTI_RC_BAD_VALUE;
+    }
+
+    LOG_DEBUG ("Sending command with TPM_CC 0x%" PRIx32 " and size %" PRIu32,
+               header.code, header.size);
+
+    rc = socket_connect (tcti_swtpm->swtpm_conf.host,
+                         tcti_swtpm->swtpm_conf.port,
+                         &tcti_swtpm->tpm_sock);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    rc = socket_xmit_buf (tcti_swtpm->tpm_sock, cmd_buf, size);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    tcti_common->state = TCTI_STATE_RECEIVE;
+
+    return rc;
+}
+
+/*
+ * In theory, we could send a cancel command to the swtpm:
+ *    tcti_control_command (tctiContext, CMD_CANCEL_TPM_CMD,
+ *                          NULL, 0, NULL, NULL, 0);
+ *
+ * However, it seems to be not implemented. A comment states:
+ *  > for cancellation to work, the TPM would have to
+ *  > execute in another thread that polls on a cancel
+ *  > flag
+ */
+TSS2_RC
+tcti_swtpm_cancel (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    (void) (tctiContext);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+tcti_swtpm_set_locality (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    uint8_t locality)
+{
+    TSS2_TCTI_SWTPM_CONTEXT *tcti_swtpm = tcti_swtpm_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_swtpm_down_cast (tcti_swtpm);
+    TSS2_RC rc;
+
+    if (tcti_swtpm == NULL) {
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+    rc = tcti_common_set_locality_checks (tcti_common);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    rc = tcti_control_command (tctiContext, CMD_SET_LOCALITY,
+                               &locality, sizeof(locality),
+                               NULL, NULL, 0);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR("Failed to set locality: 0x%" PRIx32, rc);
+        return rc;
+    }
+
+    tcti_common->locality = locality;
+    return TSS2_RC_SUCCESS;
+}
+
+TSS2_RC
+tcti_swtpm_get_poll_handles (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    TSS2_TCTI_POLL_HANDLE *handles,
+    size_t *num_handles)
+{
+    (void)(tctiContext);
+    (void)(handles);
+    (void)(num_handles);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+void
+tcti_swtpm_finalize(
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TSS2_TCTI_SWTPM_CONTEXT *tcti_swtpm = tcti_swtpm_context_cast (tctiContext);
+
+    if (tcti_swtpm == NULL) {
+        return;
+    }
+
+    socket_close (&tcti_swtpm->tpm_sock);
+}
```

Comment:
finalize function will also need to free the tcti_swtpm->conf_copy allocated in Init.

Reply:
Good call! Thanks :)